### PR TITLE
Added purging for mobileapps endpoints

### PR DESCRIPTION
**Pull request by Pchelolo**
Created a basic event service. For now it will directly purge the caches, but in future it we will emit the event to change propagation service from there.

### CODE REVIEW

File: config.test.yaml [hunk 1]
```diff
@@ -30,6 +30,9 @@ default_project: &default_project
             cache-control: s-maxage=864000, max-age=86400
           mobileapps:
             host: http://appservice.wmflabs.org
+          purge:
```
Comment (gwicke) [inline config.test.yaml]: For consistency, perhaps call this 'event'?

File: projects/wikimedia.org_sqlite.yaml [hunk 2]
```diff
@@ -68,3 +68,6 @@ paths:
       /pageviews:
         - path: sys/pageviews_proxy.yaml
           options: '{{options.pageviews}}'
+      /events:
+        - path: sys/events.js
+          options: '{{options.purge}}'
```
Comment (gwicke) [inline projects/wikimedia.org_sqlite.yaml]: .events?

File: sys/events.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    this.options = options;
+
+    if (options && options.host && options.port) {
+        this.purger = new Purger({ routes: [ options ] });
+    }
+};
+
+EventService.prototype.emitEvent = function(hyper, req) {
+    var self = this;
+    if (this.purger) {
```
Comment (gwicke) [inline sys/events.js]: Perhaps also validate the body:
```javascript
if (!Array.isArray(req.body)) {
  throw new HTTPError(...);
}
```

File: sys/events.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        self.purger.purge(req.body.map(function(event) {
+            return event.uri.toString();
```
Comment (gwicke) [inline sys/events.js]: Some validation here would be nice too.

File: sys/events.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        }))
+        .catch(function(e) {
+            hyper.log('error/purge', e);
```
Comment (gwicke) [inline sys/events.js]: error/events/purge ?

**Comment by gwicke** [comment]
Nice and clean overall!

File: sys/mobileapps.yaml [hunk 6]
```diff
@@ -51,7 +51,15 @@ paths:
...
               body: '{{from_service.body.remaining}}'
-        - return:
+        - purge_caches:
+            request:
+              method: post
+              uri: /{domain}/sys/events/
+              body:
+                - uri: //{{request.params.domain}}/api/rest_v1/page/mobile-sections/{request.params.title}
+                - uri: //{{request.params.domain}}/api/rest_v1/page/mobile-sections-lead/{request.params.title}
+                - uri: //{{request.params.domain}}/api/rest_v1/page/mobile-sections-remaining/{request.params.title}
```
Comment (gwicke) [inline sys/mobileapps.yaml]: I kind of wished we had gone with `req.` instead of `request.`, but.. too late.

Comment (Pchelolo) [inline sys/mobileapps.yaml]: @gwicke we have `request` everywhere in the code, could replace all at once later for consistency

File: sys/mobileapps.yaml [hunk 7]
```diff
@@ -51,7 +51,15 @@ paths:
               uri: /{domain}/sys/key_value/mobileapps.remaining/{title}
               headers: '{{from_service.headers}}'
               body: '{{from_service.body.remaining}}'
-        - return:
+        - purge_caches:
```
Comment (gwicke) [inline sys/mobileapps.yaml]: Minor: Perhaps call this `purge_and_return`?

File: sys/mobileapps.yaml (hunk 6 above)
Comment (gwicke) [inline sys/mobileapps.yaml]: Yeah, definitely not urgent. Just a bit of lamenting on the side.

**Comment by gwicke** [comment]
LGTM. Proper escaping depends on https://github.com/wikimedia/swagger-router/pull/38, so lets merge this in coordination with that patch.

**Comment by Pchelolo** [comment]
Note: for this to work we need to deploy a config change.

**Comment by Pchelolo** [comment]
LGTM

**Comment by d00rman** [comment]
Why require the purge body request to contain objects with only the `uri` property? It would be cleaner to just have an array with strings containing the URIs, wouldn't it?

**Comment by Pchelolo** [comment]
@d00rman The `events` module will be a general event-emitting service, se we a being general right from the beginning.

**Comment by gwicke** [comment]
Yeah, lets not make this purge-specific.

**Pull request merged**